be/src/storage/lake/compaction_policy.cpp

```diff
@@ -87,32 +87,29 @@ struct RowsetStat {
 class RowsetCandidate {
 public:
     RowsetCandidate(RowsetMetadataPtr rp, const RowsetStat& rs) : rowset_meta_ptr(std::move(rp)), stat(rs) {}
-    double calc_del_bytes() const { return (double)stat.bytes * (double)stat.num_dels / (double)stat.num_rows; }
-    // The goal of lake primary table compaction:
-    // 1. clean up deleted bytes.
-    // 2. merge rowsets with bigger compaction score
-    // 3. merge small rowsets to bigger rowset.
-    // so we pick rowset to compact by this logic:
-    // First, pick out rowset with more deleted bytes.
-    // Second, pick out rowset with bigger compaction score
-    // Finally, pick out rowset with less bytes.
-    bool operator<(const RowsetCandidate& other) const {
-        if (calc_del_bytes() < other.calc_del_bytes()) {
-            return true;
-        } else if (calc_del_bytes() > other.calc_del_bytes()) {
-            return false;
-        } else if (rowset_compaction_score() < other.rowset_compaction_score()) {
-            return true;
-        } else if (rowset_compaction_score() > other.rowset_compaction_score()) {
-            return false;
-        } else {
-            // may happen when deleted rows is zero
-            return stat.bytes > other.stat.bytes;
+    // The goal of lake primary table compaction is to reduce the overhead of reading data.
+    // So the first thing we need to do is quantify the overhead of reading the data.
+    // In object storage, we can use this to define overhead:
+    //
+    // OverHead (score) = IO count / Read bytes
+    //
+    // Same bytes, if we use more io to fetch it, that means more overhead.
+    // And in one rowset, the IO count is equal overlapped segment count plus their delvec files.
+    double io_count() const {
+        // rowset_meta_ptr->segments_size() could be zero here, so make sure this >= 1 using max.
+        double cnt = rowset_meta_ptr->overlapped() ? std::max(rowset_meta_ptr->segments_size(), 1) : 1;
+        if (stat.num_dels > 0) {
+            // if delvec file exist, that means we need to read segment files and delvec files both
+            // And update_compaction_delvec_file_io_ratio control the io amp ratio of delvec files, default is 2.
+            // Bigger update_compaction_delvec_file_io_amp_ratio means high priority about merge rowset with delvec files.
+            cnt *= config::update_compaction_delvec_file_io_amp_ratio;
         }
+        return cnt;
     }
-    double rowset_compaction_score() const {
-        return rowset_meta_ptr->overlapped() ? rowset_meta_ptr->segments_size() : 1;
-    }
+    double delete_bytes() const { return (double)stat.bytes * ((double)stat.num_dels / (double)stat.num_rows); }
+    double read_bytes() const { return (double)stat.bytes - delete_bytes() + 1; }
+    double score() const { return (io_count() * 1024 * 1024) / read_bytes(); }
+    bool operator<(const RowsetCandidate& other) const { return score() < other.score(); }
```

Comment:
Calculate the score in the constructor to avoid recalculating it every time in `<` operator

Reply:
Done